server/api/src/rsExecMyRule.cpp

```diff
@@ -1,152 +1,143 @@
+        log_api::error(msg);
         addRErrorMsg(&_comm->rError, remoteFlag, msg.data());
         return remoteFlag;
     }
 
-    if ( remoteFlag == REMOTE_HOST ) {
-        return remoteExecMyRule( _comm, _exec_inp,
-                                   _out_arr, rods_svr_host );
+    if (remoteFlag == REMOTE_HOST) {
+        return remoteExecMyRule(_comm, _exec_inp, _out_param_arr, remote_host);
     }
 
-    char* inst_name_str = getValByKey(
-                              &_exec_inp->condInput,
-                              irods::KW_CFG_INSTANCE_NAME);
+    const char* inst_name_str = getValByKey(&_exec_inp->condInput, irods::KW_CFG_INSTANCE_NAME);
     std::string inst_name;
-    if( inst_name_str ) {
+    if (inst_name_str) { // NOLINT(readability-implicit-bool-conversion)
         inst_name = inst_name_str;
     }
 
-    ruleExecInfo_t rei;
-    initReiWithDataObjInp( &rei, _comm, NULL );
-
-    // initReiWithDataObjInp allocates a KeyValPair struct. Free it now
-    // as it is immediately overwritten by the input KeyValPair.
-    free(rei.condInputData);
+    // Construct and initialize an REI object.
+    ruleExecInfo_t rei{};
+    rei.rsComm = _comm;
     rei.condInputData = &_exec_inp->condInput;
-
-    /* need to have a non zero inpParamArray for execMyRule to work */
-    if ( _exec_inp->inpParamArray == NULL ) {
-        _exec_inp->inpParamArray =
-            ( msParamArray_t * ) malloc( sizeof( msParamArray_t ) );
-        memset( _exec_inp->inpParamArray, 0, sizeof( msParamArray_t ) );
+    if (_comm) { // NOLINT(readability-implicit-bool-conversion)
+        rei.uoic = &_comm->clientUser;
+        rei.uoip = &_comm->proxyUser;
     }
-    rei.msParamArray = _exec_inp->inpParamArray;
 
-    rstrcpy( rei.ruleName, EXEC_MY_RULE_KW, NAME_LEN );
-
-    const std::string& my_rule_text   = _exec_inp->myRule;
-    const std::string& out_param_desc = _exec_inp->outParamDesc;
-    irods::rule_engine_context_manager<
-        irods::unit,
-        ruleExecInfo_t*,
-        irods::AUDIT_RULE> re_ctx_mgr(
-                               irods::re_plugin_globals->global_re_mgr,
-                               &rei);
-    irods::error err = re_ctx_mgr.exec_rule_text(
-                           inst_name,
-                           my_rule_text,
-                           _exec_inp->inpParamArray,
-                           out_param_desc);
-    if(inst_name.empty()) {
-        freeRErrorContent(&rei.rsComm->rError);
+    // Need to have a non zero inpParamArray for execMyRule to work.
+    if (!_exec_inp->inpParamArray) { // NOLINT(readability-implicit-bool-conversion)
+        // NOLINTNEXTLINE(cppcoreguidelines-owning-memory, cppcoreguidelines-no-malloc)
+        _exec_inp->inpParamArray = static_cast<MsParamArray*>(std::malloc(sizeof(MsParamArray)));
+        std::memset(_exec_inp->inpParamArray, 0, sizeof(MsParamArray));
     }
+    rei.msParamArray = _exec_inp->inpParamArray;
```

Comment:
This is just some questionable code.

This feels like the caller should be required to set `_exec_inp->inpParamArray` instead of the server doing this.

Reply:
Leaving as is. The current code does not report any leaks stemming from this code. And it's not that questionable. It seems it is more about convenience than anything else.